DelaySpec: should work as passthrough if should delay returns false

and others

https://travis-ci.com/github/akka/alpakka/jobs/364151701

```
info] Delay

[info] - should work as passthrough if should delay returns false *** FAILED *** (35 milliseconds)

[info]   java.lang.AssertionError: expected OnNext, found OnComplete

[info]   at akka.stream.testkit.TestSubscriber$ManualProbe.expectNext(StreamTestKit.scala:370)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.$anonfun$new$6(DelaySpec.scala:42)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.$anonfun$new$6$adapted(DelaySpec.scala:39)

[info]   at scala.collection.immutable.Range.foreach(Range.scala:158)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.$anonfun$new$2(DelaySpec.scala:39)

[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)

[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)

[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)

[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)

[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)

[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)

[info]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1076)

[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)

[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.withFixture(DelaySpec.scala:19)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.invokeWithFixture$1(AnyWordSpecLike.scala:1074)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTest$1(AnyWordSpecLike.scala:1086)

[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest(AnyWordSpecLike.scala:1086)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest$(AnyWordSpecLike.scala:1068)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.runTest(DelaySpec.scala:19)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTests$1(AnyWordSpecLike.scala:1145)

[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)

[info]   at scala.collection.immutable.List.foreach(List.scala:392)

[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)

[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:390)

[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:427)

[info]   at scala.collection.immutable.List.foreach(List.scala:392)

[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)

[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)

[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests(AnyWordSpecLike.scala:1145)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests$(AnyWordSpecLike.scala:1144)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.runTests(DelaySpec.scala:19)

[info]   at org.scalatest.Suite.run(Suite.scala:1112)

[info]   at org.scalatest.Suite.run$(Suite.scala:1094)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.org$scalatest$wordspec$AnyWordSpecLike$$super$run(DelaySpec.scala:19)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$run$1(AnyWordSpecLike.scala:1190)

[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.run(AnyWordSpecLike.scala:1190)

[info]   at org.scalatest.wordspec.AnyWordSpecLike.run$(AnyWordSpecLike.scala:1188)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.org$scalatest$BeforeAndAfterAll$$super$run(DelaySpec.scala:19)

[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)

[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)

[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)

[info]   at akka.stream.alpakka.googlecloud.bigquery.impl.util.DelaySpec.run(DelaySpec.scala:19)

[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:318)

[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:513)

[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:304)

[info]   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)

[info]   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)

[info]   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)

[info]   at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/Delay.scala
+++ b/google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/Delay.scala
@@ -24,7 +24,7 @@ private[impl] object Delay {
     GraphDSL.create() { implicit builder =>
       import GraphDSL.Implicits._
 
-      val splitter = builder.add(Splitter[T](shouldDelay)())
+      val splitter = builder.add(Splitter[T](shouldDelay))
       val delayFlow =
         builder.add(DelayFlow[T](() => new FibonacciStrategy[T](delayUnit, maxDelay)))
       val merge = builder.add(Merge[T](2, eagerComplete = true))

--- a/google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/Splitter.scala
+++ b/google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/Splitter.scala
@@ -6,25 +6,53 @@ package akka.stream.alpakka.googlecloud.bigquery.impl.util
 
 import akka.NotUsed
 import akka.annotation.InternalApi
-import akka.stream.scaladsl.{Broadcast, Flow, GraphDSL}
-import akka.stream.{Graph, UniformFanOutShape}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream.{Attributes, Graph, Inlet, Outlet, UniformFanOutShape}
 
 @InternalApi
 private[impl] object Splitter {
-
-  def apply[T](out0Predicate: T => Boolean)(
-      out1Predicate: T => Boolean = (elem: T) => !out0Predicate(elem)
-  ): Graph[UniformFanOutShape[T, T], NotUsed] = GraphDSL.create() { implicit builder =>
-    import GraphDSL.Implicits._
-
-    val broadcast = builder.add(Broadcast[T](2, true))
-    val filterOut0 = builder.add(Flow[T].filter(out0Predicate(_)))
-    val filterOut1 = builder.add(Flow[T].filter(out1Predicate(_)))
-
-    broadcast.out(0) ~> filterOut0
-    broadcast.out(1) ~> filterOut1
-
-    UniformFanOutShape(broadcast.in, filterOut0.out, filterOut1.out)
+  def apply[T](out0Predicate: T => Boolean): Graph[UniformFanOutShape[T, T], NotUsed] = {
+    new Splitter[T](out0Predicate)
   }
 
+  private[impl] class Splitter[T](out0Predicate: T => Boolean) extends GraphStage[UniformFanOutShape[T, T]] {
+    val in: Inlet[T] = Inlet[T]("Splitter.in")
+    val out0: Outlet[T] = Outlet("Splitter.out0")
+    val out1: Outlet[T] = Outlet("Splitter.out1")
+    override val shape: UniformFanOutShape[T, T] = UniformFanOutShape(in, out0, out1)
+
+    override def createLogic(attr: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) {
+        var pulls = 0
+        def pullIfNeeded(): Unit = {
+          if (pulls == 2) {
+            pull(in)
+          }
+        }
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            val e = grab(in)
+            if (out0Predicate(e)) {
+              push(out0, e)
+              pulls -= 1;
+            } else {
+              push(out1, e)
+              pulls -= 1;
+            }
+          }
+        })
+        setHandler(out0, new OutHandler {
+          override def onPull(): Unit = {
+            pulls += 1
+            pullIfNeeded()
+          }
+        })
+        setHandler(out1, new OutHandler {
+          override def onPull(): Unit = {
+            pulls += 1
+            pullIfNeeded()
+          }
+        })
+      }
+  }
 }

--- a/google-cloud-bigquery/src/test/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/SplitterSpec.scala
+++ b/google-cloud-bigquery/src/test/scala/akka/stream/alpakka/googlecloud/bigquery/impl/util/SplitterSpec.scala
@@ -33,7 +33,7 @@ class SplitterSpec
     RunnableGraph.fromGraph(GraphDSL.create(sink1, sink2)((_, _)) { implicit builder => (s1, s2) =>
       import GraphDSL.Implicits._
 
-      val splitter = builder.add(Splitter[Int](_ < 3)())
+      val splitter = builder.add(Splitter[Int](_ < 3))
 
       source ~> splitter.in
       splitter.out(0) ~> s1

